File: LeetCode/Python/Medium/525-Contiguous-Array/solution.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        result=0

        n=len(nums)

        prefix_sums=[0 for _ in range(0, n)]

        prefix_sums[0]=-1 if nums[0]==0 else 1

        for index in range(1, n):
            if nums[index]==1:
                prefix_sums[index]=prefix_sums[index - 1] + 1
            else:
                prefix_sums[index]=prefix_sums[index - 1] - 1

        hashmap=dict()

        for (index, prefix_sum) in enumerate(prefix_sums):
            if prefix_sum not in hashmap.keys():
                if prefix_sum==0:
                    length=index + 1

                    result=max(result, length)
                else:
                    hashmap[prefix_sum]=index
            else:
                length=index - hashmap[prefix_sum]

                result=max(result, length)

        return result

File: LeetCode/Rust/Medium/525-Contiguous-Array/solution.rs
use std::collections::{HashMap};
use std::cmp;

impl Solution
{
    pub fn find_max_length(nums: Vec<i32>) -> i32
    {
        let mut result: i32=0;

        let n: usize=nums.len();

        let mut prefix_sums: Vec<i32>=vec![0; n];

        if nums[0]==0
        {
            prefix_sums[0]-=1;
        }
        else
        {
            prefix_sums[0]+=1;
        }

        for index in 1..n
        {
            if nums[index]==1
            {
                prefix_sums[index]=prefix_sums[index - 1] + 1;
            }
            else
            {
                prefix_sums[index]=prefix_sums[index - 1] - 1;
            }
        }

        let mut hashmap: HashMap<i32, i32>=HashMap::new();

        for index in 0..n
        {
            if !hashmap.contains_key(&prefix_sums[index])
            {
                if prefix_sums[index]==0
                {
                    let length: i32=(index + 1) as i32;

                    result=std::cmp::max(result, length);
                }
                else
                {
                    hashmap.insert(prefix_sums[index], index as i32);
                }
            }
            else
            {
                let previous_index: i32=*hashmap.get(&prefix_sums[index]).unwrap();

                let length: i32=index as i32 - previous_index;

                result=std::cmp::max(result, length);
            }
        }

        result
    }
}